vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+ *
+ * For those want to extend RRGraph data structure
+ * --------------------------------------------------------------------------
+ * Please avoid modifying any existing public/private accessors/mutators
+ * in order to keep a stable RRGraph object in the framework
+ * Developers may add more internal data to RRGraph as well as associate accessors/mutators
+ * Please update and comment on the added features properly to keep this data structure friendly to be extended.
+ *
+ * Try to keep your extension within only graph-related internal data to RRGraph.
+ * In other words, extension is necessary when the new node/edge attributes are needed.
+ * RRGraph should NOT include other data which are shared by other data structures outside.
+ * Instead, using indices to point to the outside data source instead of embedding to RRGraph
+ * For example: 
+ *   For any placement/routing cost related information, try to extend t_rr_indexed_data, but not RRGraph
+ *   For any placement/routing results, try to extend PlaceContext and RoutingContext, but not RRGraph
+ * 
+ * For those want to develop placers or routers
+ * --------------------------------------------------------------------------
+ * The RRGraph is designed to be a read-only database/graph, once created.
+ * Placement and routing should NOT change any attributes of RRGraph.
+ * Any placement and routing results should be stored in other data structures, such as PlaceContext and RoutingContext. 
+ *
+ * Tracing Cross-Reference 
+ * =======================
+ * RRGraph is designed to a self-contained data structure as much as possible.
+ * It includes the switch information (rr_switch) and segment_information (rr_segment)
+ * which are necessary to build-up any external data structures.
+ *
+ * Internal cross-reference
+ * ------------------------
+ *
+ *  +--------+                  +--------+
+ *  |        |  node_in_edges   |        |
+ *  |        |  node_out_edges  |        |
+ *  |  Node  |----------------->|  Edge  |--+
+ *  |        |<-----------------|        |  |
+ *  |        |  edge_src_nodes  |        |  |
+ *  +--------+  edge_sink_nodes +--------+  |edge_switch
+ *      |                                   |
+ *      | node_segment                      |
+ *      v                                   v
+ *  +------------+                     +----------+
+ *  |            |                     |          |
+ *  |            |                     |          |
+ *  |  Segments  |                     |  Switch  |
+ *  |            |                     |          |
+ *  |            |                     |          |
+ *  +------------+                     +----------+
+ *
+ *
+ * External cross-reference
+ * ------------------------
+ * The only cross-reference to outside data structures is the cost_index
+ * corresponding to the data structure t_rr_index_data
+ * Details can be found in the definition of t_rr_index_data
+ *
```

Comment:
This allows rapid look up by the router of additional information it needs for this node, using a flyweight pattern.